rclpy/src/rclpy/node.cpp

```diff
@@ -126,46 +126,43 @@ get_node_names_impl(py::capsule pynode, bool get_enclaves)
     throw RCLError("Failed to get node names");
   }
 
-  rcutils_ret_t fini_names_ret;
-  rcutils_ret_t fini_namespaces_ret;
-  rcutils_ret_t fini_enclaves_ret;
-
   py::list pynode_names_and_namespaces(node_names.size);
-  for (size_t idx = 0; idx < node_names.size; ++idx) {
-    if (get_enclaves) {
-      pynode_names_and_namespaces[idx] = py::make_tuple(
-        py::str(node_names.data[idx]),
-        py::str(node_namespaces.data[idx]),
-        py::str(enclaves.data[idx]));
-    } else {
-      pynode_names_and_namespaces[idx] = py::make_tuple(
-        py::str(node_names.data[idx]),
-        py::str(node_namespaces.data[idx]));
-    }
-  }
 
   RCPPUTILS_SCOPE_EXIT(
     {
-      fini_names_ret = rcutils_string_array_fini(&node_names);
-      fini_namespaces_ret = rcutils_string_array_fini(&node_namespaces);
-      fini_enclaves_ret = rcutils_string_array_fini(&enclaves);
-      if (RCUTILS_RET_OK != fini_names_ret) {
+      for (size_t idx = 0; idx < node_names.size; ++idx) {
+        if (get_enclaves) {
+          pynode_names_and_namespaces[idx] = py::make_tuple(
+            py::str(node_names.data[idx]),
+            py::str(node_namespaces.data[idx]),
+            py::str(enclaves.data[idx]));
+        } else {
+          pynode_names_and_namespaces[idx] = py::make_tuple(
+            py::str(node_names.data[idx]),
+            py::str(node_namespaces.data[idx]));
+        }
+      }
+
```

Comment:
@gbalke this won't have the intended effect. If it throws, it'll throw from a destructor and C++ will terminate the program :)

If I may, you may want to do something like:

```c++
RCPPUTILS_SCOPE_EXIT({  // setup cleanup block to be run on scope exit
  rcutils_ret_t fini_ret = rcutils_string_array_fini(&node_names);
  if (RCUTILS_RET_OK != fini_ret) {
    RCUTILS_SAFE_FWRITE_TO_STDERR(
      "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
      "failed to fini node names during error handling: ");
    RCUTILS_SAFE_FWRITE_TO_STDERR(rcl_get_error_string().str);
    RCUTILS_SAFE_FWRITE_TO_STDERR("\n");
    rcl_reset_error();
  }
  fini_ret = rcutils_string_array_fini(&node_namespaces);
  if (RCUTILS_RET_OK != fini_ret) {
    RCUTILS_SAFE_FWRITE_TO_STDERR(
      "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
      "failed to fini node namespaces during error handling: ");
    RCUTILS_SAFE_FWRITE_TO_STDERR(rcl_get_error_string().str);
    RCUTILS_SAFE_FWRITE_TO_STDERR("\n");
    rcl_reset_error();
  }
  fini_ret = rcutils_string_array_fini(&enclaves);
  if (RCUTILS_RET_OK != fini_ret) {
    RCUTILS_SAFE_FWRITE_TO_STDERR(
      "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
      "failed to fini enclaves string array during error handling: ");
    RCUTILS_SAFE_FWRITE_TO_STDERR(rcl_get_error_string().str);
    RCUTILS_SAFE_FWRITE_TO_STDERR("\n");
    rcl_reset_error();
  }
});

py::list pynode_names_and_namespaces(node_names.size);
for (size_t idx = 0; idx < node_names.size; ++idx) {
  if (get_enclaves) {
    pynode_names_and_namespaces[idx] = py::make_tuple(
      py::str(node_names.data[idx]),
      py::str(node_namespaces.data[idx]),
      py::str(enclaves.data[idx]));
  } else {
    pynode_names_and_namespaces[idx] = py::make_tuple(
      py::str(node_names.data[idx]),
      py::str(node_namespaces.data[idx]));
  }
}

return pynode_names_and_namespaces;  // and then cleanup code is run automatically
```

Reply:
What I guess i don't understand is that the fini calls will happen before the for loop? Or is that the special magic of the `SCOPE_EXIT` statement? Where that will be called upon completion?

Reply:
Read the comments on the blocks and now it makes sense :+1:. Thanks for walking me through it haha.

Reply:
> Or is that the special magic of the SCOPE_EXIT statement?

If you look under the hood, this `RCPPUTILS_SCOPE_EXIT` macro instantiates an object that takes your code, wraps it in a lambda function, and executes it on _destruction_. Using object construction/destruction to manage resource lifecycles is so common in C++ that [it has a name](https://en.cppreference.com/w/cpp/language/raii#:~:text=Resource%20Acquisition%20Is%20Initialization%20or,in%20limited%20supply)%20to%20the). It's the primary idiom when it comes to exception safety.